lxd/main_nsexec.go

```diff
@@ -117,22 +90,22 @@ int dosetns(int pid, char *nstype) {
 
 void attach_userns(int pid) {
 	char nspath[PATH_MAX];
-	char userns_source[PATH_MAX];
-	char userns_target[PATH_MAX];
+	char userns_source[17];
+	char userns_target[17];
 
 	sprintf(nspath, "/proc/%d/ns/user", pid);
 	if (access(nspath, F_OK) == 0) {
-		if (readlink("/proc/self/ns/user", userns_source, 18) < 0) {
+		if (readlink("/proc/self/ns/user", userns_source, 17) < 0) {
 			fprintf(stderr, "Failed readlink of source namespace: %s\n", strerror(errno));
 			_exit(1);
 		}
 
-		if (readlink(nspath, userns_target, PATH_MAX) < 0) {
+		if (readlink(nspath, userns_target, 17) < 0) {
```

Comment:
These `readlink()` calls are buggy. `readlink()` does not `\0`-terminate. So assume that you read `5 bytes` and and there's no `\0` byte than this might give false results however unlikely. Since you don't care about `readlink()` truncating when it reads more than `17` bytes the easiest is probably to do:
```
char userns_source[17] = {0};
char userns_target[17] = {0};
```

Reply:
Ended up doing manual null termination instead (at length returned by readlink).